packages/react/src/composite/root/useCompositeRoot.ts

```diff
@@ -105,10 +105,10 @@ export function useCompositeRoot(params: UseCompositeRootParameters) {
             return;
           }
 
-          const isRtl =
-            textDirectionRef?.current == null
-              ? getTextDirection(element) === 'rtl'
-              : textDirectionRef.current === 'rtl';
+          if (textDirectionRef?.current == null) {
+            textDirectionRef.current = getTextDirection(element);
+          }
+          const isRtl = textDirectionRef.current === 'rtl';
```

Comment:
Caching this in a ref to mitigate against spamming the keys raised in https://github.com/mui/base-ui/issues/831#issuecomment-2503552760